### PR TITLE
feat/monitoring

### PR DESCRIPTION
**Pull request by kenkus-futurice**

### CODE REVIEW

File: protos/monitoring.proto [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+package proto;
+
+option java_package = "fi.hsl.common.monitoring.proto";
+option java_outer_classname = "Monitoring";
+
+message MonitoringMessage {
+    required int32 SchemaVersion = 1 [default = 1];
+    optional int64 timestamp = 2;
+    required string key = 3;
+    required double value = 4;
```
Comment (paasovaara) [inline protos/monitoring.proto]: Are we sure it's always a double? Counts are usually better as integers.. 

Can be done by having optional fields for both and perhaps a type-enumeration.. Do we wanna go that way? Of course we can always create a schema version 2.

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class MonitoringScheduler {
+    protected static final Logger LOGGER = LoggerFactory.getLogger(MonitoringScheduler.class);
+
+    protected static List<AbstractMonitoringHandler> handlers = new ArrayList<>();
+    protected static ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();;
```
Comment (paasovaara) [inline src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: extra semicolon. 😂

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    protected static long delay;
+    protected static long interval;
+    protected static Producer producer;
+    protected static Runnable runnable;
+
+    protected static MonitoringScheduler ourInstance = new MonitoringScheduler();
+
+    public static MonitoringScheduler getInstance(final PulsarApplicationContext context, final long delay, final long interval, final Runnable runnable) {
+        producer = context.getProducer();
```
Comment (paasovaara) [inline src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: Bug in the Singleton pattern here, if you call getInstance() twice with different contexts, the static producer will get overridden. Same goes for runnable. "Real" singleton pattern would store the instance into static member and then return that if once defined. One option is to make the member fields non-static, then calling getInstance is safe. 

Dunno if this is a bug or just a feature here?

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            return handlers.add(handler);
+        }
+        return false;
+    }
+
+    public boolean removeHandler(final AbstractMonitoringHandler handler) {
+        return handlers.remove(handler);
+    }
+
+    protected Runnable getRunnable() {
```
Comment (paasovaara) [inline src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: Didn't understand the role of this function without reading it completely through. Could we find a more descriptive name for it..?

File: protos/monitoring.proto (hunk 1 above)
Comment (kenkus-futurice) [reply protos/monitoring.proto]: The idea is that the value is always numeric so `double` could represent it whether it is `int` or `float` (and small errors are not a problem?).

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java (hunk 2 above)
Comment (kenkus-futurice) [reply src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: :)

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java (hunk 3 above)
Comment (kenkus-futurice) [reply src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: Indeed, perhaps this shouldn't be singleton.

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java (hunk 4 above)
Comment (kenkus-futurice) [reply src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: The idea is that this is the default task for the scheduler (get value from each "reporter" and reset the reporter, then send the value as a pulsar message), but could be overridden if needed.

File: src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                TypedMessageBuilder<byte[]> messageBuilder = producer.newMessage()
+                        .eventTime(timestamp)
+                        .value(message.toByteArray());
+                messageBuilder.sendAsync()
+                        .whenComplete((id, throwable) -> {
+                            if (throwable != null) {
+                                LOGGER.warn("Failed to send Pulsar message.", throwable);
+                            }
+                        });
+            });
```
Comment (paasovaara) [inline src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: Java 8 stream().forEach will swallow all exceptions (It's a real pain in the ass..). Would be nice to do something on errors, at least log them?

Comment (kenkus-futurice) [reply src/main/java/fi/hsl/common/monitoring/MonitoringScheduler.java]: shiit, good to know.

**Comment by kenkus-futurice** [comment]
This is probably not needed anymore as monitoring is provided externally.

**Comment by mjaakko** [comment]
Closing this pull request as monitoring is done externally

**Pull request closed without merging**